jack-core/src/com/rapleaf/jack/DatabaseConnection.java

```diff
@@ -101,21 +102,38 @@ public DatabaseConnection(String dbname_key, long expiration, String driverClass
    * Get a Connection to a database. If there is no connection, create a new one.
    * If the connection hasn't been used in a long time, close it and create a new one.
    * We do this because MySQL has an 8 hour idle connection timeout.
+   * <p>
+   * Because of the intermittent downtime of SQL service, it implements exponential retry policy.
+   * The default retry policy retries seven times, which handles SQL downtime less than approx. two minutes.
    */
 
   public Connection getConnectionInternal() {
-    try {
-      if (conn == null) {
-        Class.forName(driverClass);
-        conn = DriverManager.getConnection(connectionString, username.orNull(), password.orNull());
-      } else if (isExpired() || conn.isClosed()) {
-        resetConnection();
+    for (int retryCount = 0; retryCount < GET_CONNECTION_MAX_RETRIES; ++retryCount) {
+      try {
+        if (conn == null) {
+          Class.forName(driverClass);
+          conn = DriverManager.getConnection(connectionString, username.orNull(), password.orNull());
+        } else if (isExpired() || conn.isClosed()) {
+          resetConnection();
+        }
+        updateExpiration();
+        return conn;
+      } catch (Exception e) { //IOEx., SQLEx.
+        // if it is the last retry, throw exception
+        if (retryCount == GET_CONNECTION_MAX_RETRIES - 1) {
+          throw new RuntimeException(String.format("Could not establish connection after %d attempts",
+              GET_CONNECTION_MAX_RETRIES), e);
+        }
+        // Wait (2^retryCount) seconds
+        try {
+          Thread.sleep(Math.round(Math.pow(2, retryCount)) * 1000);
+        } catch (Exception ignored) {
+          // Catches interruptions in the sleep
```

Comment:
Can you store this exception and include it as part of the final `RuntimeException` you throw? Otherwise, we will know it failed to create a connection will be known but we won't know why.

Reply:
The program is not supposed to reach final `RuntimeException` because it should either successfully create a connection or fail seven times and throw a `RuntimeException` in the 7th try, which includes the detail of the `exception`. 
In this case, do we still need to store this exception?

p.s. The try-catch block of `Thread.sleep` is to ensure the retry continues if the process got interrupted during sleep, and I didn't use the exception since it doesn't contain any interesting information;

Reply:
Oh, sorry. I completely misread.

Reply:
nvm I think it is better to include the information in the final `runtimeException` in case it reaches to the final one.